homeassistant/components/xiaomi_miio/gateway.py

```diff
@@ -115,10 +116,14 @@ def connect_gateway(self):
 
             try:
                 miio_cloud = MiCloud(self._cloud_username, self._cloud_password)
-                if not miio_cloud.login():
-                    raise ConfigEntryAuthFailed(
-                        "Could not login to Xiaomi Miio Cloud, check the credentials"
-                    )
+                for attempt in range(4):
+                    if miio_cloud.login():
```

Comment:
Ideally, we make a change upstream so we know the difference between a failed login or a connection error.

IN case of a connection error, we can throw a ConfigEntryNotReady, while in case of an authentication failure we can throw a ConfigEntryAuthFailed.

That way, the retry logic becomes a Home Assistant problem.

Reply:
I made an upstream PR to diffrentiate AuthFailed from Socket Errors: https://github.com/Squachen/micloud/pull/2

Reply:
the upsteam PR has been merged and used in this PR